src/components/ble/AppleNotificationCenterClient.cpp

```diff
@@ -0,0 +1,140 @@
+}
+
+int NewAlertSubcribeCallback(uint16_t conn_handle, const struct ble_gatt_error* error, struct ble_gatt_attr* attr, void* arg) {
+  auto client = static_cast<AppleNotificationCenterClient*>(arg);
+  return client->OnNewAlertSubcribe(conn_handle, error, attr);
+}
+
+AppleNotificationCenterClient::AppleNotificationCenterClient(Pinetime::System::SystemTask& systemTask,
+                                                 Pinetime::Controllers::NotificationManager& notificationManager)
+  : systemTask {systemTask}, notificationManager {notificationManager} {
+}
+
+bool AppleNotificationCenterClient::OnDiscoveryEvent(uint16_t connectionHandle, const ble_gatt_error* error, const ble_gatt_svc* service) {
+  if (service == nullptr && error->status == BLE_HS_EDONE) {
+    if (isDiscovered) {
+      NRF_LOG_INFO("ANCS Discovery found, starting characteristics discovery");
+
+      ble_gattc_disc_all_chrs(connectionHandle, ancsStartHandle, ancsEndHandle, OnANCSCharacteristicDiscoveredCallback, this);
+    } else {
+      NRF_LOG_INFO("ANCS not found");
+      onServiceDiscovered(connectionHandle);
+    }
+    return true;
+  }
+
+  if (service != nullptr && ble_uuid_cmp(&ancsUuid.u, &service->uuid.u) == 0) {
+    NRF_LOG_INFO("ANCS discovered : 0x%x - 0x%x", service->start_handle, service->end_handle);
+    ancsStartHandle = service->start_handle;
+    ancsEndHandle = service->end_handle;
+    isDiscovered = true;
+  }
+  return false;
+}
+
+int AppleNotificationCenterClient::OnCharacteristicsDiscoveryEvent(uint16_t connectionHandle, const ble_gatt_error* error, const ble_gatt_chr* characteristic) {
+  if (error->status != 0 && error->status != BLE_HS_EDONE) {
+    NRF_LOG_INFO("ANCS Characteristic discovery ERROR");
+    onServiceDiscovered(connectionHandle);
+    return 0;
+  }
+
+  if (characteristic == nullptr && error->status == BLE_HS_EDONE) {
+    NRF_LOG_INFO("ANCS Characteristic discovery complete");
+    if (isCharacteristicDiscovered) {
+      ble_gattc_disc_all_dscs(connectionHandle, notificationSourceHandle, ancsEndHandle, OnANCSDescriptorDiscoveryEventCallback, this);
+    } else {
+      onServiceDiscovered(connectionHandle);
+    }
+  } else {
+    if (characteristic != nullptr && ble_uuid_cmp(&notificationSourceChar.u, &characteristic->uuid.u) == 0) {
+      NRF_LOG_INFO("ANCS Characteristic discovered: Notification Source");
+      notificationSourceHandle = characteristic->val_handle;
+      // notificationSourceDefHandle = characteristic->def_handle;
+      isCharacteristicDiscovered = true;
+    }
+  }
+  return 0;
+}
+
+int AppleNotificationCenterClient::OnDescriptorDiscoveryEventCallback(uint16_t connectionHandle, const ble_gatt_error* error, uint16_t characteristicValueHandle, const ble_gatt_dsc* descriptor) {
+  if (error->status == 0) {
+    if (characteristicValueHandle == notificationSourceHandle && ble_uuid_cmp(&notificationSourceChar.u, &descriptor->uuid.u)) {
```

Comment:
Why does the last comparison needed?
This callback is called only for notificationSourceChar's descriptors, no?

Reply:
I tried to recall this. This handles the notification subscription, right? Then we also want this for the dataSourceChar as well, but this code as it is just tries to detect an incoming message and then to show a dummy "Hello World" Notification. So this should stay in, because later on, there will be another case.